src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/CacheAspectTests.cs

```diff
@@ -1208,228 +1242,136 @@ public void NullableAndDefaultConstructors_TwoParams_Nullable()
 
-            VerifyAllMocks(CacheService);
+            Assert.Equal($"Cache key type cannot be null", exception.Message);
         }
 
         [Fact]
-        public void TwoCaches_CacheHitOnSecond()
+        public void InvalidKeyType()
         {
-            var cacheKey1 = new TestCacheKeyWithString("test");
-            var cacheKey2 = new TestCacheKeyWithInt(10);
-
-            var expected = new TestValue();
-
-            CacheService
-                .Setup(s => s.GetItem(cacheKey1, typeof(TestValue)))
-                .ReturnsAsync(null);
-            CacheService
-                .Setup(s => s.GetItem(cacheKey2, typeof(TestValue)))
-                .ReturnsAsync(expected);
-
-            var result = TestMethods.TwoCaches("test", 10);
-
-            Assert.Equal(expected, result);
+            var exception = Assert.Throws<ArgumentException>(TestMethods.InvalidKeyType);
 
-            VerifyAllMocks(CacheService);
+            Assert.Equal(
+                $"Cache key type {typeof(object).FullName} must be assignable to {typeof(ICacheKey).GetPrettyFullName()}",
+                exception.Message
+            );
         }
 
         [Fact]
-        public void TwoCaches_CacheMissAll()
+        public void ForceUpdate()
         {
-            var cacheKey1 = new TestCacheKeyWithString("test");
-            var cacheKey2 = new TestCacheKeyWithInt(10);
-
-            CacheService
-                .Setup(s => s.GetItem(cacheKey1, typeof(TestValue)))
-                .ReturnsAsync(null);
-            CacheService
-                .Setup(s => s.GetItem(cacheKey2, typeof(TestValue)))
-                .ReturnsAsync(null);
+            var cacheKey = new TestCacheKey();
 
             var args = new List<object>();
 
+            // Verify that there is no attempt to "get" a currently-cached value when the "ForceUpdate" flag is set.
+            // This means that we're only ever getting fresh values for the item and then setting or updating the cached
+            // entry with that new value rather than ever attempting to retrieve it.
             CacheService
-                .Setup(s => s.SetItem(cacheKey1, Capture.In(args)))
-                .Returns(Task.CompletedTask);
-            CacheService
-                .Setup(s => s.SetItem(cacheKey2, Capture.In(args)))
-                .Returns(Task.CompletedTask);
+                .Setup(s => s.SetItem(cacheKey, Capture.In(args)));
 
-            var result = TestMethods.TwoCaches("test", 10);
+            var returnedItem = TestMethods.ForceUpdate();
 
-            Assert.Equal(args[0], result);
-            Assert.Equal(args[1], result);
+            var cachedItem = Assert.Single(args);
+            Assert.Equal(cachedItem, returnedItem);
 
             VerifyAllMocks(CacheService);
         }
 
-        [Fact]
-        public void TwoCaches_CustomOrder_CacheHitOnFirst()
+        private static void AssertCacheHit(ICacheKey cacheKey, object expectedResult, Func<TestValue> run)
```

Comment:
It'd be good if we could implement some unit or integration tests that verify that the previous deadlocking behaviour is no longer occurring.

If we make any more changes to the caching code in the future, it'd be good to know that we aren't introducing any regressions. Getting feedback at this stage would be much better than relying on the performance tests (esp. as we don't run them all the time).

I think @benoutram was leaning more towards integration tests as they probably give a more accurate reflection of how this would work in production.

Not entirely sure what we would need to do here, but would suggest we split out a separate task for this as it could get pretty gnarly.